src/base/LruMap.h

```diff
@@ -14,21 +14,34 @@
 #include <list>
 #include <map>
 
-template <class EntryValue, size_t EntryCost = sizeof(EntryValue)> class LruMap
+template<class E>
+size_t
+DefaultCostSize(const E *e)
+{
+    return sizeof(*e);
+}
+
+template <class EntryValue, size_t EntryCost(const EntryValue *) = DefaultCostSize>
```

Comment:
The original bad design combined with this change opens a small Pandora box: Since our `add()` method stores and our `get()` method returns non-constant EntryValues, the cost of a cached EntryValue may change. For example, a stored SSL context may gain more certificates and "grow" in size (we currently hard-code context sizes but that is a known bug, not a design guideline). We now need to document how entry cost changes are handled (at the very least).

The implementation implies that entry sizes are computed twice: Once when the entry is cached (naturally) and once when the entry is purged from the cache. The code reader is left to wonder whether the two computed cost values must always match. It is also difficult to tell whether runtime changes in a cached entry size may lead to the now-too-big entry being purged from the cache.

N.B. The old constant entry cost was not realistic but it did not raise these questions, of course.

If we document what has been implemented so far, we will arrive at something like this:

> Since `add()`, the EntryCost function must return the same value for the same EntryValue pointer, even if the size of the underlying entry value has changed.

Needless to say, that requirement is very onerous in general. We should probably change the implementation to avoid placing such an unreasonable requirement on class users. There are many ways to close this Pandora box, but I will discuss just three:

**A**: Remove the second template parameter. Add a memory usage/cost parameter to `add()`. Store the supplied memory usage/cost value in Entry. Always use the stored value, ignoring any possible EntryValue size changes. The `memUsed_` member is just a rough estimate (for LruMaps with size-changing entries): We may cache a lot more or a lot less than we are allowed to.

**B**: Same changes as in approach A plus replace `get()` with a `find()` method that returns (a pointer to) a _constant_ entry value and also add a checkout()/checkin() interface that deals with access to non-constant values. The checkin() method will accept size. Monitor for size changes during checkout (and purge entries to honor the configured cache size limit). Prohibit entry value changes outside of checkout. The `memUsed_` member continues to reflect true memory usage.

**C**: Same changes as in approach B but instead of a checkout()/checkin() interface, add an `extract()` interface that purges the entry from the cache. If the caller needs to add the entry back, they start from scratch with `add()`. The `memUsed_` member continues to reflect true memory usage.

I recommend B or C because they do not lie about memory usage. Approach B is more efficient because its checkout/checkin sequence does not remove and then rehash the previously cached entry (unlike the extract/add sequence in approach C).

What do you think?